contrib/gorilla/mux/mux.go

```diff
@@ -3,125 +3,59 @@ package mux
+// The last parameter is optional and allows to pass a custom tracer.
+func NewRouter(service string, trc ...*tracer.Tracer) *Router {
+	t := getTracer(trc)
+	t.SetServiceInfo(service, "gorilla/mux", ext.AppTypeWeb)
+	return &Router{mux.NewRouter(), t, service}
+}
+
+// ServeHTTP dispatches the request to the handler
+// whose pattern most closely matches the request URL.
+// We only need to rewrite this function to be able to trace
+// all the incoming requests to the underlying multiplexer
+func (r *Router) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	var match mux.RouteMatch
+	var route string
+	var err error
+
+	// get the resource associated to this request
+	if r.Match(req, &match) {
+		route, err = match.Route.GetPathTemplate()
+		if err != nil {
+			route = "unkown"
 		}
-
-		// trace the request
-		tracedRequest, span := m.trace(req)
-		defer span.Finish()
-
-		// trace the response
-		tracedWriter := newTracedResponseWriter(span, writer)
-
-		// run the request
-		handler(tracedWriter, tracedRequest)
-	}
-}
-
-// HandleFunc will add a traced version of the given handler to the router.
-func (m *MuxTracer) HandleFunc(router *mux.Router, pattern string, handler http.HandlerFunc) *mux.Route {
-	return router.HandleFunc(pattern, m.TraceHandleFunc(handler))
-}
-
-// span will create a span for the given request.
-func (m *MuxTracer) trace(req *http.Request) (*http.Request, *tracer.Span) {
-	route := mux.CurrentRoute(req)
-	path, err := route.GetPathTemplate()
-	if err != nil {
-		// when route doesn't define a path
-		path = "unknown"
-	}
-
-	resource := req.Method + " " + path
-
-	span := m.tracer.NewRootSpan("mux.request", m.service, resource)
-	span.Type = ext.HTTPType
-	span.SetMeta(ext.HTTPMethod, req.Method)
-	span.SetMeta(ext.HTTPURL, path)
-
-	// patch the span onto the request context.
-	treq := SetRequestSpan(req, span)
-	return treq, span
-}
-
-// tracedResponseWriter is a small wrapper around an http response writer that will
-// intercept and store the status of a request.
-type tracedResponseWriter struct {
-	span   *tracer.Span
-	w      http.ResponseWriter
-	status int
-}
-
-func newTracedResponseWriter(span *tracer.Span, w http.ResponseWriter) *tracedResponseWriter {
-	return &tracedResponseWriter{
-		span: span,
-		w:    w}
-}
-
-func (t *tracedResponseWriter) Header() http.Header {
-	return t.w.Header()
-}
-
-func (t *tracedResponseWriter) Write(b []byte) (int, error) {
-	if t.status == 0 {
-		t.WriteHeader(http.StatusOK)
+	} else {
+		route = "unkown"
 	}
-	return t.w.Write(b)
-}
+	resource := req.Method + " " + route
 
-func (t *tracedResponseWriter) WriteHeader(status int) {
-	t.w.WriteHeader(status)
-	t.status = status
-	t.span.SetMeta(ext.HTTPCode, strconv.Itoa(status))
-	if status >= 500 && status < 600 {
-		t.span.Error = 1
-	}
+	// we need to wrap the ServeHTTP method to be able to trace it
+	httptrace.Trace(r.Router.ServeHTTP, w, req, r.service, resource, r.Tracer)
```

Comment:
I guess we can just do:
```
httptrace.Trace(r.Router, w, req, resource, r)
```
And then `Trace` can get both `service` and `Tracer` from `Router`.

Reply:
Or would the original form be preferred?